crates/noirc_evaluator/src/ssa/context.rs

```diff
@@ -424,6 +471,190 @@ impl<'a> SsaContext<'a> {
         self[target_block].instructions.insert(1, phi_id);
         phi_id
     }
+
+    fn memcpy(&mut self, l_type: ObjectType, r_type: ObjectType) {
+        if l_type == r_type {
+            return;
+        }
+
+        if let (ObjectType::Pointer(a), ObjectType::Pointer(b)) = (l_type, r_type) {
+            let len = self.mem.arrays[a as usize].len;
+            let adr_a = self.mem.arrays[a as usize].adr;
+            let adr_b = self.mem.arrays[b as usize].adr;
+            for i in 0..len {
+                let idx_b = self.get_or_create_const(
+                    FieldElement::from((adr_b + i) as i128),
+                    ObjectType::Unsigned(32),
+                );
+                let idx_a = self.get_or_create_const(
+                    FieldElement::from((adr_a + i) as i128),
+                    ObjectType::Unsigned(32),
+                );
+                let load = self.new_instruction(idx_b, idx_b, Operation::Load(b), r_type);
+                self.new_instruction(load, idx_a, Operation::Store(a), l_type);
+            }
+        } else {
+            dbg!(&l_type);
+            dbg!(&r_type);
+            unreachable!("invalid type, expected arrays");
+        }
+    }
+
+    pub fn handle_assign(&mut self, lhs: NodeId, index: Option<NodeId>, rhs: NodeId) -> NodeId {
+        let lhs_type = self.get_object_type(lhs);
+        let rhs_type = self.get_object_type(rhs);
+        let i = 0;
+        let i_obj =
+            self.get_or_create_const(FieldElement::from(i as i128), ObjectType::Unsigned(32));
+        if let Some(Instruction {
+            operator: Operation::Call(_), ins_arguments: call_stack, ..
+        }) = self.try_get_mut_instruction(rhs)
+        {
+            if index.is_some() || !matches!(lhs_type, ObjectType::Pointer(_)) {
+                let obj_type = if let ObjectType::Pointer(a) = lhs_type {
+                    self.mem.arrays[a as usize].element_type
+                } else {
+                    lhs_type
+                };
+                let ret = self.new_instruction(rhs, i_obj, Operation::Res, obj_type);
+                return self.handle_assign(lhs, index, ret);
+            } else {
+                call_stack.return_values.push(lhs);
+                return lhs;
+            }
+        }
+        if let Some(idx) = index {
+            if let ObjectType::Pointer(a) = lhs_type {
+                //Store
+                return self.new_instruction(
+                    rhs,
+                    idx,
+                    Operation::Store(a),
+                    self.mem.arrays[a as usize].element_type,
+                );
+            } else {
+                unreachable!("Index expression must be for an array");
+            }
+        } else if matches!(lhs_type, ObjectType::Pointer(_)) {
+            if let Some(Instruction {
+                operator: Operation::Intrinsic(_), res_type: rtype, ..
```

Comment:
Having a complex function is not a bad thing per se. On the contrary, it is much better if the complex business logic is gathered in one point rather than being scattered into multiple places, which is what I believe I have done.
That being said it is indeed good practice to split big functions into smaller pieces, but I don't think the size of this one is such a big issue.

Reply:
I rather disagree that complex functions are not bad. Complex functions are perhaps the easiest way to hide bugs since bugs are easier to spot in smaller functions that do fewer things. Breaking up a larger function into smaller chunks also makes the individual steps easier to understand since you'll likely end up calling a function for each step.